### ISSUE TITLE
Get-Item P:\ath\To\A\Directory -Streams * does not work

### ISSUE THREAD

**Issue by n3rdopolis** [PowerShell/PowerShell#10570]
I tested this in PowerShell 5.1, and PowerShell 7 preview 3.

If I add an Alternate Data Stream to a directory, verify with streams.exe that it exists, and then try to list the ADS stream for the directory itself with PowerShell
```
Get-Item P:\ath\To\A\Directory -Streams * 
```
does not work, and appears to return nothing

**Comment by iSazonov** [comment]
How do you add a stream to the directory?  Please add more info in repo steps.

**Comment by vexx32** [comment]
Are wildcards supported for that parameter? I don't think that works for files with multiple streams either.

**Comment by iSazonov** [comment]
Wildcards is supported
https://github.com/PowerShell/PowerShell/blob/79e80a8e06ddf286b231dc853347059f6da37d39/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L1253-L1280

**Comment by n3rdopolis** [comment]
You can do in a command prompt
echo hi > C:\Downloads:test

or 
set-output "test" C:\downloads -stream test
in PowerShell to create a stream.


Some folders like $ENV:TEMP I get access denied oddly.

**Comment by iSazonov** [comment]
All cmdlets (Get-Item, Set-Content, Get-Content, Clear-Content) work only with alternative streams on files. Code contains explicit checks for this.

@SteveL-MSFT We need a conclusion should we support alternative streams on directories.

**Comment by n3rdopolis** [comment]
Support on it would be nice, to be able to write scripts that get directory sizes. Although probably very rare, someone could make a huge file as an ADS on a directory.

**Comment by SteveL-MSFT** [comment]
I don't believe this requires @PowerShell/powershell-committee review.  If sysinternals streams.exe supports it meaning NTFS supports it, then there's no reason not to have PowerShell support it.  However, it wouldn't be a priority as this is the first time this has come up and it's not clear to me what use case would add an alternate stream to a folder.

**Comment by n3rdopolis** [comment]
Whoo! Now for the record, I'm not looking to add them myself to folders, I don't see the use case for adding ADS streams to folders as well, but the script I made is for retrieving how large sub folders are. And I want to be able to count that, in case it actually happens...

**Comment by iSazonov** [comment]
We could add a support for NTFS streams for directories at FileSystem provider V2 time.

**Comment by kyanha** [comment]
Not going to lie, this issue is why I downloaded the code.  I'm looking at a filesystem filter driver that needs to make use of alternate data streams, and I need to ensure I have tools to look at them to find and ferret out bugs.

That said, since it's well-known that Powershell doesn't allow administrators to even see alternate data streams on directories, they provide a very good place for a red team to store data for exfiltration.  The fact that Powershell doesn't support alternate stream syntax on directories is arguably a bug in Powershell.  (As has been noted, NTFS itself supports alternate data streams on directories, and has supported them since Windows NT 3.1.)

This is an incredibly easy patch.  (System.Management.Automation\namespaces\FileSystemProvider.cs, Microsoft.Powershell.Commands.FileSystemProvider.GetItem(string path). Just get rid of the isContainer check.)  It took me about 3 hours to find where to fix it, and about 2 minutes to understand how to fix it.

I have built a local pwsh that doesn't have the isContainer check, and I have tested that it works with Get-Item.  However, a potential issue is raised due to the nature of NTFS alternate data streams on files versus directories:

1. A file always has at least one stream of type $DATA, which is unnamed.  This means that `Get-Item filen.ame -Stream *` will always return at least one entry on files stored on NTFS.
2. A directory usually has zero streams of type $DATA.  This means that `Get-Item dirn.ame -Stream *` will usually return nothing, which leads (interactively) to Get-Item returning a red message of `Get-Item: Reached the end of the file.`.

Put simply, "a directory is a file which can contain zero data streams."  I haven't tested the behavior of powershell-as-distributed on FAT32 or exFAT, where neither files nor directories contain alternate data streams, but I suspect that the `Reached the end of the file.` message would be returned in all cases on those filesystems (either that or an exception would be thrown).

So, I have a question: should this functionality be enabled by an additional flag to the Get-Item command (like "-EnableDirectoryStreams" or such)?  Or should the isContainer check simply be removed, since containers in the NTFS filesystem can have streams and should be able to be seen regardless?

**Comment by vexx32** [comment]
If that's tried on a non-NTFS filesystem, does the cmdlet throw a reasonable error? If we're going to enable it, we need to ensure that any errors arising from its use on non-NTFS file systems are handled in a reasonably clear fashion.

But yeah, if the given file system supports it I'm not sure there's really a reason to not allow it.

**Comment by kyanha** [comment]
> If that's tried on a non-NTFS filesystem, does the cmdlet throw a reasonable error? If we're going to enable it, we need to ensure that any errors arising from its use on non-NTFS file systems are handled in a reasonably clear fashion.

In powershell-as-distributed, if you attempt to get streams from a file stored on a FAT32 filesystem, it returns `Get-Item: The parameter is incorrect.`  If you attempt to get streams from a directory stored on a FAT32 filesystem, it returns nothing (the same way as it works on NTFS, because it simply doesn't try to get the streams from anything that passes the isContainer check).

In my local build, the `Get-Item: The parameter is incorrect.` message shows up on both files and directories stored on FAT32, as expected.

**Comment by vexx32** [comment]
Yeah, we should probably make sure to catch that exception and wrap it so we can give a bit more useful of an error message, like, perhaps something to the effect of `The stream <name | *> could not be accessed for the item <Path>. The filesystem reported the following error: <Error text from the filesystem error>`

**Comment by kyanha** [comment]
erm, is there an example of where that's being done elsewhere in the codebase?  I have no qualms with it, I just don't know how to do it in a way that the code will work with.  (I mean, what's the exception to construct and throw?  Do I include the text of the caught exception myself, or is that something that the parent handler can do by accessing my thrown exception's InnerException property?)

**Comment by kyanha** [comment]
@vexx32 I uploaded a WIP of the (extremely minor) change to let -stream work on directories, but I'm not sure the best options for wrapping the exception (as seen in my prior comment).  I'm more than happy to work on it, and if you'd like to add to it please feel free.

**Comment by vexx32** [comment]
Okay, this isn't something I'm super familiar with, but from a glance at the code I'd probably bet the exception initially gets thrown from the call to `GetStreams()`

In your current build of PowerShell, you'll need to trigger the exception and then check `$error[0].Exception.GetType().FullName` -- that's probably the exception type you should target in the code.

I would probably recommend moving the `GetStreams()` call to before loop and store the result in a variable; that way, you can just have the try/catch around the GetStreams() call itself and not the entire loop.

Something similar to this pattern would most likely be effective:

```csharp
try
{
    var alternateStreams = AlternateDataStreamUtilities.GetStreams(result.FullName);
}
catch (Exception e) // replace Exception with the specific exception type you found earlier
{
    // you can probably reuse the WriteError() code from after the foreach loop here
    // though you'll need to locate the error string in the `*.resx` file (the name should match the class name, "FileSystemProviderStrings")
    // and add a new format string to reference so you can add in the `e.Message` from the caught exception here
    WriteError( ... );

    // finally just `return` from the method since we've already written an error and know there are no streams
    // that we can use for the foreach loop
    return;
}


foreach (AlternateStreamData stream in alternateStreams) {
    // ... (rest of the existing code in the loop etc)
```
Hopefully that makes some sense.

**Comment by kyanha** [comment]
https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirststreamw says that the "End of file has been reached." (ERROR_HANDLE_EOF) is an expected return value when the file in question has no ::$DATA streams.  If I catch that, would it make sense to only WriteError it if the isContainer is false?  That would assuage a concern that I have, that running `Get-Item * -stream *` might create spurious error stacks due to directories in the current directory not having any data streams.

**Comment by vexx32** [comment]
Maybe. I'm not really sure, it depends on how the `AlternateDataStreamUtilities.GetStreams()` API used here handles that error; that's a native error, so I'd generally hope that most common errors like that are handled by the API in use and wrapped in proper .NET exceptions before they leave the API.

**Comment by iSazonov** [comment]
I think we should follow common behavior for globbing - no exceptions/errors if glob char presents `Get-Item -Stream*` otherwise write an terminating error.

/cc @mklement0

**Comment by vexx32** [comment]
I don't think a terminating exception is a good idea; IIRC it's possible to do `Get-Item *.txt -Streams $streamName` so the error here should always be non-terminating IMO. Best not to overcomplicate the logic.

**Comment by kyanha** [comment]
> Maybe. I'm not really sure, it depends on how the `AlternateDataStreamUtilities.GetStreams()` API used here handles that error; that's a native error, so I'd generally hope that most common errors like that are handled by the API in use and wrapped in proper .NET exceptions before they leave the API. So I'd be thinking that if that's the case, the API should throw a different kind of exception; that's why I recommended you check the exception type and use that to catch it, so that other kinds of exceptions still show through normally, if they were going to be reported.

AlternateDataStreamUtilities.GetStreams() is a direct usage of the Win32 API via P/Invoke.  (There exists no alternate data stream API in the .NET BCL.)  The ERROR_HANDLE_EOF ("End of file has been reached.") return was a direct result of a `throw new Win32Exception()`.

The other error return from that function that I've seen is ERROR_INVALID_PARAMETER, which is what you get from filesystems that don't implement data streams.  It is also returned as a thrown Win32Exception().

**Comment by kyanha** [comment]
Looking at the semantics here (for testing, from both current behavior and from discussion):

Current code makes a nonexistent specificstream a terminating exception.  I have not changed this behavior.

file:specificstream exists, no exception.
file:specificstream nonexists, terminating exception.
file:wildcardstream exists, no exception.
file:wildcardstream nonexists, no exception.
directory:specificstream exists, no exception.
directory:specificstream nonexists, terminating exception.
directory:wildcardstream exists, no exception.
directory:wildcardstream nonexists, no exception.

vexx32 thinks that it should be a nonterminating exception.
iSazonov thinks that it should be a terminating exception.

@vexx32 What's the rationale for changing current code?  Do you believe such a change to this behavior could pass the backward compatibility contract test under bucket 2 or bucket 3?

@iSazonov What's the rationale for keeping it a terminating exception?  Does the behavior parity concern apply to a glob application across multiple searches? (If you do a `Get-Item *.txt -Stream specificstream` it essentially repeats the search on every *.txt file, whereas the original glob of *.txt runs a single search against the current directory. Should a single file that is matched from the original glob but which doesn't contain specificstream prevent the search for specificstream from continuing across every requested file?)

I will say that my own use case would be better served by it being a nonterminating exception, and I think that it makes more sense for a search failure on one file to not terminate the search against all other files.  That said, I'm not even a contributor here yet, and what I think doesn't really have as much weight as yours.

Also, I think that any change to the behavior for multiple files will necessitate committee review.  I'd like to get your thoughts before raising the issue for them to consider.

**Comment by vexx32** [comment]
If that's the existing behaviour, we should probably stick with it then. I might have my own opinions, but changing the existing behaviour is probably not a good idea at the moment. We can revisit it after the functionality has been added in if we feel it's needed, but I think keeping the added functionality separate from any breaks in backwards compatibility is a good idea when we're able to do that. 🙂

**Comment by kyanha** [comment]
I concur.

**Comment by iSazonov** [comment]
> iSazonov thinks that it should be a terminating exception.

I was wrong. Non-terminating error is ok.

**Comment by mklement0** [comment]
I've only glanced at the discussion, so forgive me if this isn't helpful, but it does look like targeting a specific (non-wildcard) stream is currently a _non-terminating_ error:

```powershell
PS> Get-Item $profile, $profile -Stream NoSuch
Get-Item: Could not open the alternate data stream 'NoSuch' of the file 'C:\Users\jdoe\Documents\PowerShell\Microsoft.PowerShell_profile.ps1'.
Get-Item: Could not open the alternate data stream 'NoSuch' of the file 'C:\Users\jdoe\Documents\PowerShell\Microsoft.PowerShell_profile.ps1'.

**Comment by iSazonov** [comment]
Thinking more I believe:
- `Get-Item -Path * -Stream *` shouldn't write errors
- `Get-Item -Path * -Stream "exact_name_without_wildcards"` should write non-terminating errors if the stream is not found.

**Comment by kyanha** [comment]
...clearly I need a refresher on the code difference between terminating and non-terminating errors.  Is there a document on it somewhere, or do I need to read more code?

**Comment by iSazonov** [comment]
@kyanha Use WriteError() for non-terminating errors.

**Comment by vexx32** [comment]
The code difference is quite small; basically `WriteError(someErrorRecord)` is non-terminating, and `ThrowTerminatingError(someErrorRecord)` is terminating.

**Comment by mklement0** [comment]
Here's some conceptual background: https://stackoverflow.com/a/39949027/45375

**Comment by kyanha** [comment]
So, if we catch an ERROR_INVALID_PARAMETER Win32Exception (which usually indicates that the filesystem doesn't support streams), would that be a case to throw a terminating error rather than a non-terminating error?  Or is it reasonable to continue on the hypothesis that we may be running on a filesystem that may support streams for some objects but not all?  (NTFS is not the only filesystem which supports filestream syntax, and there's no enforcement within the kernel that NTFS's stream semantics are implemented or adhered to by the filesystem.)

Put another way, how tightly do we want to couple the Win32 PowerShell with NTFS?

**Comment by vexx32** [comment]
I think if the filesystem reports invalid parameter it seems reasonable to throw a terminating error, because that indicates it doesn't support alternate streams, right?

Or does that same Win32 error indicate more than just that?

**Comment by iSazonov** [comment]
> I think if the filesystem reports invalid parameter it seems reasonable to throw a terminating error, because that indicates it doesn't support alternate streams, right?

I think it is another case. We can ignore this in current time and wait a feedback.

**Comment by iSazonov** [comment]
> Put another way, how tightly do we want to couple the Win32 PowerShell with NTFS?

I believe we should think only about NTFS, ReFS and *FAT on Windows. All other file systems is "edge case" and we should wait feedbacks.

**Comment by mklement0** [comment]
As a general rule, terminating errors should only be thrown _if the cmdlet cannot continue to process further inputs_ (see the previous SO link), which is not the case here, especially given that you can hypothetically pass multiple paths to `Get-Item` that refer to different file systems.

**Comment by vexx32** [comment]
Ah, that's a very good point. Non-terminating makes the most sense then, aye.

**Comment by kyanha** [comment]
Hey, I'm having some trouble with writing the tests, because I don't really know what's happening here:

```ps
    Context "Alternate Stream Tests" {
        BeforeAll {
            if ( $skipNotWindows )
            {
                return
            }
            $altStreamPath = "$TESTDRIVE/altStream.txt"
			$altStreamDirectory = "$TESTDRIVE/altstreamdir"
            $stringData = "test data"
            $streamName = "test"
			$absentStreamName = "noExist"
            $item = New-Item -type file $altStreamPath
            Set-Content -Path $altStreamPath -Stream $streamName -Value $stringData
			$diritem = New-Item -type directory $altStreamDirectory
			Set-Content -Path $altStreamDirectory -Stream $streamName -Value $stringData
        }
        It "Should find an alternate stream on a file if present" -Skip:$skipNotWindows {
            $result = Get-Item $altStreamPath -Stream $streamName
            $result.Length | Should -Be ($stringData.Length + [Environment]::NewLine.Length)
            $result.Stream | Should -Be $streamName
        }
		It "Should not find an alternate stream on a file if not present" -Skip:$skipNotWindows {
			$result = Get-Item $altStreamPath -Stream $absentStreamName
			$result.Length | Should -Be (0)
		}
		It "Should find an alternate stream on a directory if present" -Skip:$skipNotWindows {
			$result = Get-Item $altStreamDirectory -Stream $streamName
			$result.Length | Should -Be ($stringData.Length + [Environment]::NewLine.Length)
			$result.Stream | Should -Be $streamName
		}
		It "Should not find an alternate stream on a directory if not present" -Skip:$skipNotWindows {
			$result = Get-Item $altStreamPath -Stream $absentStreamName
			$result.Length | should -Be (0)
		}
    }
```
Gives this error:
```
    Context Alternate Stream Tests
      [-] Error occurred in Context block 0ms
        NotSupportedException: Unable to clear content of 'D:\Users\kyanha\AppData\Local\Temp\39e98245-6db5-49a4-ba1e-02a96e91a90c\altstreamdir' because it is a directory. Clear-Content is only supported on files.
        at <ScriptBlock>, D:\Users\kyanha\work\github\Powershell\Powershell\test\powershell\Modules\Microsoft.PowerShell.Management\Get-Item.Tests.ps1: line 122
        at Invoke-Blocks, D:\Users\kyanha\work\github\Powershell\Powershell\src\powershell-win-core\bin\Debug\net5.0\win7-x64\publish\Modules\Pester\4.10.1\Functions\SetupTeardown.ps1: line 135
        at Invoke-TestGroupSetupBlocks, D:\Users\kyanha\work\github\Powershell\Powershell\src\powershell-win-core\bin\Debug\net5.0\win7-x64\publish\Modules\Pester\4.10.1\Functions\SetupTeardown.ps1: line 121
        at DescribeImpl, D:\Users\kyanha\work\github\Powershell\Powershell\src\powershell-win-core\bin\Debug\net5.0\win7-x64\publish\Modules\Pester\4.10.1\Functions\Describe.ps1: line 209
```
Line 122 is `Set-Content -Path $altStreamDirectory -Stream $streamName -Value $stringData`.

I haven't been able to figure out how to run my local pwsh build under a debugger (I usually use Visual Studio and I can't build it from there, and I haven't figured out how to start it under its debugger) so I can't really trace through it.

**Comment by vexx32** [comment]
```
Set-Content -Path $altStreamDirectory -Stream $streamName -Value $stringData
```

Looks like Set-Content is hard-coded to not allow writing to directories, alternate streams or no... can you use Set-Item for that purpose here instead? If not, there might be more needed to allow setting alternate streams on directories :/

**Comment by kyanha** [comment]
The exception text refers to Clear-Content, not Set-Content.  (The original PR was to read alternate streams, not write them, though I would not be averse to writing them as well.)  Will anyone get mad at me if I execute commands with `cmd.exe /c` in the test setup, to at least ensure the environment for reading gets set up correctly?

**Comment by vexx32** [comment]
Yeah, I think Set-Content internally makes a call to the Clear-Content api first. That's my best guess since Clear-Content isn't called directly there.

And yeah that probably should be a separate PR, so using cmd /c calls is fine if need be... assuming those tests are only run on Windows, which I think is the case?

**Comment by kyanha** [comment]
Yeah, the entire set is headed by 
```ps
if ( $skipNotWindows ) {
    return
}
```
None of the AlternateDataStream tests are run on any platform other than Windows.

**Comment by kyanha** [comment]
Patch updated with tests.  Can someone doublecheck to make sure I did them correctly, please?
